### PR TITLE
Add tutorial for PDF reader

### PR DESCRIPTION
**Pull request by 20001LastOrder**
# Your checklist for this pull request
Thank you for submitting a pull request! To speed up the review process, please follow this checklist:

- [ ] My Pull Request is small and focused on one topic so it can be reviewed easily 
- [ ] My code follows the style guidelines of this project (`make format`)
- [ ] Commit messages are detailed
- [ ] I have performed a self-review of my code
- [ ] I commented hard-to-understand parts of my code
- [ ] I updated the documentation (docstrings, `/docs`)
- [ ] My changes generate no new warnings (or explain any new warnings and why they're ok)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] All tests pass when I run `pytest tests` (offline mode)

Additional steps for code with networking dependencies:

- [ ] I followed the [offline and online testing guidelines](https://sherpa-ai.readthedocs.io/en/latest/How_To/Test/test.html#offline-and-online-testing)
- [ ] All tests pass when I run `pytest tests --external_api` (online mode, making network calls)

# Description
Describe your pull request here.

What does this PR implement or fix? Explain.

If this PR resolves any currently open issues then mention them like this: `Closes #xxx`.
Github will close such issues automatically when your PR is merged into `main`.

Any relevant logs, error output, etc?

Any other comments? For example, will other contributors need to install new libraries via `poetry install` after picking up these changes?

💔Thank you!

### CODE REVIEW

File: docs/How_To/Tutorials/document_reader.rst
Comment (amirfz) [reply docs/How_To/Tutorials/document_reader.rst:22]: sigh! this took me a long time to fix. had to reinstall the whole anaconda. it kept giving me a circular error that I needed to upgrade conda first

Comment (amirfz) [reply docs/How_To/Tutorials/document_reader.rst:22]: maybe we create a separate issue for a docker file that contains the right env?

File: docs/How_To/Tutorials/document_reader.rst
Comment (amirfz) [inline docs/How_To/Tutorials/document_reader.rst:31]: I need to install LangChain and loguru too, no?

File: docs/How_To/Tutorials/document_reader.rst
Comment (amirfz) [inline docs/How_To/Tutorials/document_reader.rst:183]: do I need to pip install these? hydra? argparse? etc

File: docs/How_To/Tutorials/document_reader.rst
Comment (amirfz) [inline docs/How_To/Tutorials/document_reader.rst:246]: this image is recorded on a very while screen, it makes it hard to read it. could you replace it with one from a narrower screen to improve readability?

File: docs/How_To/Tutorials/document_reader.rst
Comment (amirfz) [inline docs/How_To/Tutorials/document_reader.rst:242]: ok, this worked nicely! I'm impressed how easy it was

File: docs/How_To/Tutorials/document_reader.rst
Comment (amirfz) [inline docs/How_To/Tutorials/document_reader.rst:111]: this is nice, but as we discussed I'm hoping for the simplest non-trivial example. can we add another action like google search? then we would have two actions which makes it nontrivial

File: docs/How_To/Tutorials/document_reader.rst
Comment (amirfz) [inline docs/How_To/Tutorials/document_reader.rst:116]: in the spirit of making this less trivial, could we also add a verification step?

**Review by amirfz — CHANGES_REQUESTED** [review]

File: docs/How_To/Tutorials/document_reader.rst
Comment (20001LastOrder) [reply docs/How_To/Tutorials/document_reader.rst:31]: These dependencies are installed at same time as sharpa_ai is installed.

File: docs/How_To/Tutorials/document_reader.rst
Comment (20001LastOrder) [reply docs/How_To/Tutorials/document_reader.rst:183]: No, these are install as part of the sherpa_ai installation

File: docs/How_To/Tutorials/document_reader.rst
Comment (20001LastOrder) [reply docs/How_To/Tutorials/document_reader.rst:111]: I added a new section to the tutorial about Google search and the verification steps

**Review by oshoma — CHANGES_REQUESTED** [review]
I added comments for text changes.

Also running into a few problems here. See following transcript from the end of the tutorial after adding the validation. 

Issues:
1. citation_validation is dying with a bug
2. The hugging face warnings make the tutorial harder to understand. Can we get rid of these? I've seen them at other times and wondered if we are doing something wrong, but haven't dug in.
3. Repeated warnings "Number of requested results 20 is greater than number of elements in index 3, updating n_results = 3" . I'm not sure where these are coming from or whether they are benign.

```
Ask me a question: what is the PDF about?
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
2024-04-08 11:59:47.833 | INFO     | sherpa_ai.agents.base:run:69 - Action selected: ('DocumentSearch', {'query': 'PDF about'})
Number of requested results 20 is greater than number of elements in index 3, updating n_results = 3
2024-04-08 11:59:49.418 | INFO     | sherpa_ai.agents.base:run:69 - Action selected: ('DocumentSearch', {'query': 'PDF about'})
Number of requested results 20 is greater than number of elements in index 3, updating n_results = 3
Traceback (most recent call last):
  File "/Users/oshoma/dev/sherpa-tutorial/main.py", line 35, in <module>
    result = qa_agent.run()
             ^^^^^^^^^^^^^^
  File "/Users/oshoma/miniconda3/lib/python3.11/site-packages/sherpa_ai/agents/base.py", line 100, in run
    result = self.validate_output()
             ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/oshoma/miniconda3/lib/python3.11/site-packages/sherpa_ai/agents/base.py", line 115, in validate_output
    validation_result = validation.process_output(result, self.belief)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/oshoma/miniconda3/lib/python3.11/site-packages/sherpa_ai/output_parsers/citation_validation.py", line 189, in process_output
    resources = self.resources_from_belief(belief)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/oshoma/miniconda3/lib/python3.11/site-packages/sherpa_ai/output_parsers/citation_validation.py", line 161, in resources_from_belief
    resources.extend(action.meta[-1])
                     ~~~~~~~~~~~^^^^
```

File: docs/How_To/Tutorials/document_reader.rst
Comment (artificialfintelligence) [inline docs/How_To/Tutorials/document_reader.rst:175]: This renders as a hyperlink. It should be formatted as `_target_`

**Comment by artificialfintelligence** [comment]
@oshoma @20001LastOrder  Some initial comments:

1. `Step 2. Create a folder for storing your tutorial code and PDF files.` and then further down under “Define the custom action”: `Create a folder for this tutorial. In this folder, create a file called actions.py and add the following code.` It sounds a bit confusing because it seemed we already created a folder "for the tutorial". Maybe change the second "folder" to "subfolder".
2. In the agent_config for qa_agent: `description: You are a Question answering assistant helping users to find answers to the text.` I think this should be “…find answers to **to their questions**” instead of “to the text”? Or maybe “**from** the text” was meant? (This works as is, but still “find answers to the text” sounds strange.)

And just a few minor issues:
1a) Under "Define the custom action", below the code, in the line below, "**name**" should also be enclosed in backticks:  
**name** and `args`: These properties describe the action to agents that want to use it.

1b) Also under "Defining the agent configuration", below the code, in the line below, “name” should also be enclosed in backticks:  
**name** and `description`: These are used to describe the agent when it is executing the task.

2\) In the line `Finally, to view more detailed logs, you can set the log level to debug by changing the LOG_LEVEL environment variable to the .env file`, this should be “**in** the .env file”

3\) `…first try to collection relevant information…` should be “… to **collect** relevant information”

4\) In the sentence "The \_target\_ keys tell Sherpa which classes to use...", `_target_` is rendered as a hyperlink in ReStructured Text. It should be written as \`\_target\_\` (with backticks).

**Comment by artificialfintelligence** [comment]
No further comments. I was able to follow this tutorial and implement this in a fresh venv without hitting any roadblocks. Very well-written and easy to follow. I just encountered a known issue with citation validation at the very end.

**Comment by amirfz** [comment]
@20001LastOrder since the citation validation util is throwing an error most of the time re #306 can we use a different verification util as part of the tutorial that is more reliable (at least until the bug is fixed)? if not, let's remove that part of the tutorial for now since the tutorial should be focused on what works robustly.

**Comment by 20001LastOrder** [comment]
I think I addressed all the comments. Waiting for approval from @oshoma and @amirfz to merge.

**Review by amirfz — DISMISSED** [review]

**Comment by oshoma** [comment]
I created these for the open issues:

- #339 
- #340

**Review by oshoma — APPROVED** [review]

**Pull request merged**
